frontend/server/k8s-helper.ts

```diff
@@ -124,6 +134,102 @@ export function getPodLogs(podName: string): Promise<string> {
   return (k8sV1Client.readNamespacedPodLog(podName, namespace, 'main') as any)
     .then(
       (response: any) => (response && response.body) ? response.body.toString() : '',
-      (error: any) => { throw new Error(JSON.stringify(error.body)); }
+      (error: any) => {throw new Error(JSON.stringify(error.body));}
     );
 }
+
+function getNameOfViewerResource(logdir: string): string {
+  return 'viewer-' + crypto.SHA1(logdir);
+}
+
+/**
+* Create Tensorboard pod via CRD with the given logdir if there is no existing
+* Tensorboard pod.
+*/
+export async function newTensorboardInstance(logdir: string): Promise<void> {
+  if (!k8sV1CustomObjectClient) {
+    throw new Error('Cannot access kubernetes Custom Object API');
+  }
+  const currentPod = await getTensorboardInstance(logdir);
+  if (currentPod) {
+    return;
+  }
+
+  // TODO: take the configuration below to a separate file
+  // Name of the viewer resource is based on logDir.
+  const body = {
+    apiVersion: viewerGroup + '/' + viewerVersion,
+    kind: 'Viewer',
+    metadata: {
+      name: getNameOfViewerResource(logdir),
+      namespace: namespace,
+    },
+    spec: {
+      type: 'tensorboard',
+      tensorboardSpec: {
+        logDir: logdir,
+      },
+      PodTemplateSpec: {
+        spec: {
+          containers: [{
+            env: [{
+              name: 'GOOGLE_APPLICATION_CREDENTIALS',
+              value: '/secret/gcp-credentials/user-gcp-sa.json'
+            }],
+            volumeMounts: [{
+              mountPath: '/secret/gcp-credentials',
+              name: 'gcp-credentials',
+            }],
+          }],
+          volumes: [{
+            name: 'gcp-credentials',
+            volumeSource: {
+              secret: {
+                secretName: 'user-gcp-sa',
+              }
+            },
+          }],
+        }
+      }
+    }
+  };
+  await k8sV1CustomObjectClient.createNamespacedCustomObject(viewerGroup,
+    viewerVersion, namespace, viewerPlural, body);
+}
+
+/**
+* Finds a running Tensorboard pod created via CRD with the give logdir and
+* returns its pod IP and port.
+*/
+export async function getTensorboardInstance(logdir: string): Promise<string> {
+  if (!k8sV1CustomObjectClient) {
+    throw new Error('Cannot access kubernetes Custom Object API');
+  }
+
+  const viewers = (await k8sV1CustomObjectClient.listNamespacedCustomObject(
```

Comment:
does client lib has getNamespacedCustomObject?
if so can we use that instead?

Reply:
done